DiverseSelector/base.py

```diff
@@ -24,71 +24,123 @@
+
 from .utils import feature_generator, feature_reader, get_features
 
+class Selector(ABC):
+    @abstractmethod
+    def pick_initial_compounds(self):
+        """Pick the initial compounds.
+        """
+        pass
+
+    @abstractmethod
+    def compute_diversity(self, metric: str, func=None):
+        """
+        Compute the diversity between every component in dataset
+        :param metric: one of the standard metrics ("Euclidean", "Cosine", "Jaccard" etc. )
+        :param func: callable function that take 2 arguments and calculate distance between them. Specified by user
+        :return: np.ndarray aka (dis-)similarity matrix
+        """
+        pass
+
+    @abstractmethod
+    def select(self, num_selected: int, random_seed: int = None):
+        """
+        Select the subset molecules with optimal diversity.
+        :param num_selected: amount of molecule that need to be selected
+        :param random_seed: random seed for random selection compounds
+        :return: np.array of indices of selected molecules from the initial dataset ## Maybe not indices....
+        """
+        pass
 
-class DissimilaritySelectionBase(ABC):
+    @abstractmethod
+    def subset_diversity(self, subset: np.ndarray, div_metric: str):
+        """
+        Calculate diversity of the subset.
+        :param subset: numpy array of indices which make up a subset
+        :param div_metric: metric for calculating diversity of the subset ("Gini", "Entropy" etc.)
+        :return: float #Scale should be discussed (0 to 1, or 0 to 100 or -1 to 1 or anything else)
+        """
+        # todo: need to implement diversity measurement here
+        pass
+
+    @abstractmethod
+    @property
+    def all_diversity(self, div_metric: str): # Not an abstract method!!! Needs to be implemented explicitly here
+        """
+        Calculate diversity of the original dataset.
+        :param div_metric: metric for calculating diversity of the subset ("Gini", "Entropy" etc.)
+        :return: float #Scale should be discussed (0 to 1, or 0 to 100 or -1 to 1 or anything else)
+        """
+        # todo: need to implement diversity measurement here
+        pass
+
+    def read_excel(self, fname: str):  # Not an abstract method!!! Needs to be implemented explicitly here
+        """
+        Read dataset from an excel file
+        :param fname: file name for saving molecule
+        :return: None
+        """
+        pass
+
+    def to_excel(self, fname: str):  # Not an abstract method!!! Needs to be implemented explicitly here
+        """
+        Save selected molecules as an excel file
+        :param fname: file name for saving molecule
+        :return: None
+        """
+        pass
+
+    def to_txt(self, fname: str): # Not an abstract method!!! Needs to be implemented explicitly here
+        """
+        Save selected molecules as an excel file
+        :param fname: file name for saving molecule
+        :return: None
+        """
+        pass
```

Comment:
Thanks for the contribution! @RichRick1 
Maybe better to put this part to `utils.py`. But please don't worry and I will make a change soon.

Reply:
I don't mind to have it as function. I've tried to mock the way that pandas work: no matter what dataset is, you can save it using method to_csv, to_excel etc. So in our case we can have the _concrete_ method that is written in abstract class and there is no need to have separate function that work with instance of a class. Instead, we assume that every object has the same structure so it can be saved in the same way _without_ calling outside function

Reply:
Now I get what you mean. I will slightly change this behavior, like adding support of exporting SDF file. Thanks for the clarification.